lib/hls/hls_parser.js

```diff
@@ -3616,6 +3620,129 @@ shaka.hls.HlsParser = class {
     return [startByte, endByte];
   }
 
+  /**
+   * @param {!Array.<!shaka.hls.Tag>} tags
+   * @private
+   */
+  processDateRangeTags_(tags) {
+    if (this.initialProgramDateTime_ == null) {
+      // We cannot process EXT-X-DATERANGE tags without
+      // EXT-X-PROGRAM-DATE-TIME.
+      return;
+    }
+    let dateRangeTags =
+        shaka.hls.Utils.filterTagsByName(tags, 'EXT-X-DATERANGE');
+    dateRangeTags = dateRangeTags.sort((a, b) => {
+      const aStartDateValue = a.getRequiredAttrValue('START-DATE');
+      const bStartDateValue = b.getRequiredAttrValue('START-DATE');
+      if (aStartDateValue < bStartDateValue) {
+        return -1;
+      }
+      if (aStartDateValue > bStartDateValue) {
+        return 1;
+      }
+      return 0;
+    });
+    for (let i = 0; i < dateRangeTags.length; i++) {
+      const tag = dateRangeTags[i];
+      const id = tag.getRequiredAttrValue('ID');
+      if (this.dateRangeIdsEmitted_.has(id)) {
+        continue;
+      }
+      const startDateValue = tag.getRequiredAttrValue('START-DATE');
+      const startDate = shaka.util.TXml.parseDate(startDateValue);
+      if (isNaN(startDate)) {
+        // Invalid START-DATE
+        continue;
+      }
+      goog.asserts.assert(startDate != null, 'Start date should not be null!');
+      const startTime = Math.max(0, startDate - this.initialProgramDateTime_);
```

Comment:
This isn't quite right... We had a massive conversation about this in video-dev#hls a while back, involving Rob Walch from HLS.js, but unfortunately I can't find it due to the messages disappearing after 90 days.

Happily, it was already the 2nd time we discussed it there so we learned form the first time, and so to keep things more persistent we moved some of the conversation into his-interest mailing group. Here is the last answer from that thread: https://mailarchive.ietf.org/arch/msg/hls-interest/CaOoXRDa1AMPClq57qVIV82vJqo/

In short, the safest anchor is the PDT who's date is closest to this date range START-DATE (but not greater), and then use the sync time to the segment to compute the offset for the `startTime`. This avoids any errors due to re-syncs of PDT around discontinuities. In fact, the HLS spec was updated after that conversation in the latest draft 15: https://author-tools.ietf.org/iddiff?url1=draft-pantos-hls-rfc8216bis-14&url2=draft-pantos-hls-rfc8216bis-15&difftype=--html

That method has issues though, as the user may have set `this.config_.hls.ignoresManifestProgramDateTime` to `true`... With that in mind I would say that you shouldn't be able to `processDateRangeTags_` if `ignoresManifestProgramDateTime` is `true` (similar to how it won't be done if `this.initialProgramDateTime_ == null`.

Reply:
If `ignoresManifestProgramDateTime` is true, you're right we should avoid processing all this, I'll add it. Regarding the use of PROGRAM-DATE-TIME, it is a general problem throughout the parser and the use that is given to it, so we will address this in another PR.

Reply:
Done!

Reply:
Thanks!

In case it may be helpful for the future PR, some of the conversation we had in video-dev / HLS Interest inspired some of the changes in this hls.js PR: https://github.com/video-dev/hls.js/pull/6213

Not sure how applicable it is here but just wanted to provide what info I have.

Thanks again for maintaining this awesome project!